scripts/data/gen-nodetime/src/relayer/lib/xrelayer.ts

```diff
@@ -0,0 +1,490 @@
+export const getPathMethod = "getPath";
+export const listPathsMethod = "listPaths";
+export const getDefaultAccountMethod = "getDefaultAccount";
+export const getDefaultAccountBalanceMethod = "getDefaultAccountBalance";
+export const linkMethod = "link";
+export const startMethod = "start";
+
+type EnsureChainSetupResponse = {
+	// id is the chain id of chain.
+	id: string;
+};
+type Account = {
+	address: string;
+};
+type ChainConfig = {
+	chainId: string;
+	rpcAddr: string;
+	addrPrefix: string;
+	gasPrice: string;
+};
+type ConnectOptions = {
+	sourcePort: string;
+	sourceVersion: string;
+	targetPort: string;
+	targetVersion: string;
+	ordering: "ORDER_UNORDERED" | "ORDER_ORDERED";
+};
+type Connections = {
+	srcConnection: string;
+	destConnection: string;
+};
+type Endpoint = {
+	channelID?: string;
+	chainID: string;
+	portID: string;
+};
+type Path = {
+	id: string;
+	isLinked: boolean;
+	src: Endpoint;
+	dst: Endpoint;
+};
+type LinkResponse = {
+	linkedPaths: string[];
+	alreadyLinkedPaths: string[];
+};
+type StartResponse = {};
+type PacketHeights = {
+	packetHeightA: number;
+	packetHeightB: number;
+	ackHeightA: number;
+	ackHeightB: number;
+};
+type PathConfig = {
+	path: Path;
+	options?: ConnectOptions;
+	connections?: Connections;
+	relayerData?: PacketHeights;
+};
+type RelayerConfig = {
+	mnemonic: string;
+	chains?: Array<ChainConfig>;
+	paths?: Array<PathConfig>;
+};
+
+type LogMethod = (message: string) => Logger;
+
+interface Logger {
+	error: LogMethod;
+	warn: LogMethod;
+	info: LogMethod;
+	verbose: LogMethod;
+	debug: LogMethod;
+}
+
+class ConsoleLogger {
+	public readonly error: LogMethod;
+	public readonly warn: LogMethod;
+	public readonly info: LogMethod;
+	public readonly verbose: LogMethod;
+	public readonly debug: LogMethod;
+
+	constructor() {
+		this.error = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.warn = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.info = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.verbose = (msg) => {
+			console.log(msg);
+			return this;
+		};
+		this.debug = (msg) => {
+			console.log(msg);
+			return this;
+		};
+	}
+}
+
+export default class XRelayer {
+	private _config: RelayerConfig;
+	public config: RelayerConfig;
+	public homedir: string;
+	private relayers: Map<string, ReturnType<typeof setInterval>>;
+
+	constructor() {
+		this.homedir = os.homedir();
+		this.relayers = new Map();
+		this._config = this.readOrCreateConfig();
+		const nestedProxy = {
+			set: (target, prop, value) => {
+				target[prop] = value;
+				this.writeConfig(this._config);
+				return true;
+			},
+			get: (target, prop) => {
+				if (typeof target[prop] === "object" && target[prop] !== null) {
+					return new Proxy(target[prop], nestedProxy);
+				} else {
+					return target[prop];
+				}
+			},
+		};
+		this.config = new Proxy(this._config, nestedProxy);
+	}
+	createConfigFolder() {
+		const configPath = this.homedir + "/.ts-relayer";
```

Comment:
I haven't refactored this yet because I was contemplating possibly passing the config file as an option through the `starport relayer` command and having it as a constructor param. What do you think? Or should we refactor this as is and revisit later?

Reply:
I think we can refactor this now because even if make the config configurable we still need this new method and the detault relayer dir constant.